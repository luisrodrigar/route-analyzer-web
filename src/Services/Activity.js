
import {Exception} from './Exception'
import axios from 'axios';

const DIR_BASE = "https://routeanalyzer-api.herokuapp.com/";

export function get(id){
	const path = DIR_BASE + "/activity/" + id ;
	return axios
		.get(path)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception("Not match any activity with the id passed as a parameter.");
    	});
}

export function exportAs(id, type){
	const path = DIR_BASE + "/activity/" + id + "/export/" + type ;
	return axios
		.get(path)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception("Problem trying to export the file as " + type);
    	});
}

export function joinLaps(id, dataSelected){
	let indexLap1 = dataSelected[0].indexLap - 1;
	let indexLap2 = dataSelected[1].indexLap - 1;

	const url = DIR_BASE + "/activity/" + id +"/join/laps?index1=" 
		+ indexLap1 + "&index2=" + indexLap2;

	return axios
		.put(url)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception(err.response.data.description);
    	});
}

export function removePoint(id, position, timeInMillis, index){
	const positionParam = getPositionParams(position);
	const time = !timeInMillis || isNaN(timeInMillis) ? null : "timeInMillis=" + timeInMillis ;
	const indexParam = "index="+index;
	
	const url = DIR_BASE + "/activity/" + id +"/remove/point?" 
		+ positionParam + (time ? "&" + time : "&timeInMillis=") + "&" + indexParam;

	return axios
		.put(url)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception(err.response.data.description);
    	});
}

export function splitLap(id, position, timeInMillis, index){
	const positionParam = getPositionParams(position);
	const time = !timeInMillis || isNaN(timeInMillis) ? null : "timeInMillis=" + timeInMillis ;
	const indexParam = "index="+index;
	
	const url = DIR_BASE + "/activity/" + id +"/split/lap?" + positionParam + "&" + indexParam + (time ? "&" + time : "&timeInMillis=");

	return axios
		.put(url)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception(err.response.data.description);
    	});
}

export function removeLaps(id,dataSelected){

	const indexLaps = dataSelected.map(lap => {
		return lap.indexLap;
	}).join(",");

	const startTimeLaps = dataSelected.map(lap => {
		return lap.startTime?lap.startTime:null;
	}).join(",");

	const url = DIR_BASE + "/activity/" + id +"/remove/laps?date=" + startTimeLaps + "&index=" + indexLaps;

	return axios
		.put(url)
    	.then(res => res.data)
    	.catch(err => {
    		throw new Exception(err.response.data.description);
    	});
}

export function setColors(id, data){
	const url = DIR_BASE + "/activity/" + id +"/color/laps?data=" + data;
	return axios
		.put(url)
		.then()
    	.catch(err => {
    		throw new Exception("It cannot be posible to set the colors");
    	});
}

function getPositionParams(position){
	return "lat="+position.lat+"&lng="+position.lng;
}