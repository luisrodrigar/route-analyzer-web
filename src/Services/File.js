import axios from 'axios';
import Exception from '../Exception'

export function uploadFile(formData) {
   this.path = "http://localhost:8080/RouteAnalyzer/file/upload";
    axios
      .post(this.path, formData)
      .then(res => { 
        return res.data[0]; 
      } )
      .catch(err =>{
        throw new Exception(err.response.data.description);
      });
}



export default File;