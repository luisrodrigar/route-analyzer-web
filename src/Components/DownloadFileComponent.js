import React, {Component} from 'react';
import {get} from '../Services/file';
import {exportAs} from '../Services/activity';
import { connect } from "react-redux";
import FileDownload from 'js-file-download';

const mapStateToProps = state => {
  return { 
    id: 	state.container.activity.id,
    type: 	state.container.activity.sourceXmlType
  };
};

export class DownloadFileComponent extends Component {

	constructor(props){
	    super(props);
	    this.download = this.download.bind(this);
  	}

	download() {
		get(this.props.id, this.props.type)
			.then(response => {
				FileDownload(response, this.props.id + "_" + this.props.type + '.xml');
			})
			.catch(err =>{
				alert(err.message);
			});
  	}

  	export(type) {
	    exportAs(this.props.id, type)
	    .then(response => {
			FileDownload(response, this.props.id + "_" + type + '.xml');
	    })
	    .catch(err => {
	      	alert(err.message);
	    });
	}
  	
	render() {
	   	return(
	   		<div>
		   		<button onClick={this.download}>Download file </button>
		   		<button onClick={()=>this.export('tcx')}>Export as TCX</button>
		   		<button onClick={()=>this.export('gpx')}>Export as GPX</button>
	   		</div>
	 	);
	}
}



export default connect(mapStateToProps)(DownloadFileComponent);