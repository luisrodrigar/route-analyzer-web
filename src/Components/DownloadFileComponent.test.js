import toJson from "enzyme-to-json";

jest.mock('js-file-download');
jest.mock("../Services/file");
jest.mock("../Services/activity");
import React from "react";
import { DownloadFileComponent } from "./DownloadFileComponent";
import { shallow } from 'enzyme';
import jsFileDownload from 'js-file-download';
import * as fileService from '../Services/file';
import * as activityService from '../Services/activity';
import flushPromises from "flush-promises";

describe("Download File Component Test", () => {
    const gpxType = "gpx";
    const tcxType = "tcx";
    let id;
    let gpxProps;
    let tcxProps;
    beforeAll(() => {
        id = "5ace8cd14c147400048aa6b0";
        gpxProps = {
            id,
            type: gpxType
        };
        tcxProps = {
            id,
            type: tcxType
        };
    });
    afterEach(() => {
      jest.resetAllMocks();
    });
    it("Initial render without crashing", async () => {
        const component = shallow(<DownloadFileComponent {...tcxProps} />);
        expect(toJson(component)).toMatchSnapshot();
    });
    it("Download a file", async () => {
        const response = {
            result: true,
            value: "Mocked response."
        };
        fileService.get.mockResolvedValueOnce(response);
        const component = shallow(<DownloadFileComponent {...gpxProps} />);
        const instance = component.instance();
        await instance.busy;
        const buttonDownload = component.find('button').at(0);
        expect(buttonDownload.text()).toBe("Download file ");
        await buttonDownload.simulate('click');
        expect(fileService.get).toBeCalledWith(id, gpxType);
        expect(jsFileDownload).toBeCalledWith(response, id + "_" + gpxType + ".xml");
    });
    it("Error downloading gpx file", async () => {
        const error = {
            message: "Mocked error response."
        };
        fileService.get.mockRejectedValueOnce(error);
        const component = shallow(<DownloadFileComponent {...gpxProps} />);
        const instance = component.instance();
        await instance.busy;
        const buttonDownload = component.find('button').at(0);
        expect(buttonDownload.text()).toBe("Download file ");
        await buttonDownload.simulate('click');
        expect(fileService.get).toHaveBeenCalledWith(id, gpxType);
        expect(jsFileDownload).not.toHaveBeenCalled();
    });
    it("Export activity to tcx file", async () => {
        const response = {
            message: "Activity exported."
        };
        activityService.exportAs.mockResolvedValueOnce(response);
        const component = shallow(<DownloadFileComponent {...tcxProps} />);
        const instance = component.instance();
        await instance.busy;
        const buttonExportTcx = component.find('button').at(1);
        expect(buttonExportTcx.text()).toBe("Export as TCX");
        await buttonExportTcx.simulate('click');
        expect(activityService.exportAs).toHaveBeenCalledWith(id, tcxType);
        expect(jsFileDownload).toHaveBeenCalledWith(response, id + "_" + tcxType + ".xml");
    });
    it("Export activity to gpx file", async () => {
        const response = {
            message: "Activity exported."
        };
        activityService.exportAs.mockResolvedValueOnce(response);
        const component = shallow(<DownloadFileComponent {...gpxProps} />);
        const instance = component.instance();
        await instance.busy;
        const buttonExportGpx = component.find('button').at(2);
        expect(buttonExportGpx.text()).toBe("Export as GPX");
        await buttonExportGpx.simulate('click');
        expect(activityService.exportAs).toHaveBeenCalledWith(id, gpxType);
        expect(jsFileDownload).toHaveBeenCalledWith(response, id + "_" + gpxType + ".xml");
    });
    it("Error exporting activity", async () => {
        window.alert = jest.fn(() => ({}));
        const error = {
            message: "Error exporting an activity."
        };
        activityService.exportAs.mockRejectedValueOnce(error);
        const component = shallow(<DownloadFileComponent {...tcxProps} />);
        const instance = component.instance();
        await instance.busy;
        const buttonExportTcx = component.find('button').at(1);
        buttonExportTcx.simulate('click');

        await flushPromises();

        expect(activityService.exportAs).toHaveBeenCalledWith(id, tcxType);
        expect(jsFileDownload).not.toHaveBeenCalledWith();
        expect(window.alert).toBeCalledWith(error.message);
    });
});

