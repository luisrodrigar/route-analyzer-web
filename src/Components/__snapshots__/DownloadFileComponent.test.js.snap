// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download File Component Test Initial render without crashing 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Download file 
  </button>
  <button
    onClick={[Function]}
  >
    Export as TCX
  </button>
  <button
    onClick={[Function]}
  >
    Export as GPX
  </button>
</div>
`;
